File: source/uController.pas
unit uController;

interface

uses
  SysUtils, uCommonDevice, uModbus;

type
  TAr = array[0..59] of byte;

  TController = class(TCommonModbusDevice)
  private
    function ReadData(Addr: byte; RegAddr: word; var Ans: word): TModbusError; overload;
    function ReadData(Addr: byte; RegAddr: word; var Ans: ansistring; Count: byte = 1): TModbusError; overload;
    function WriteState(Addr: byte; RegAddr: word; Value: array of word): TModbusError;
  public
    //Данные сканера входа
    function GetEnterCard(Addr: byte; var IsNew: boolean; var CardNum: TAr; var Count: word): TModbusError;
    //Данные сканера выхода
    function GetExitCard(Addr: byte; var IsNew: boolean; var CardNum: TAr; var Count: word): TModbusError;
    //Состояние входа
    //0	– проход закрыт;
    //(1-$FFFЕ) – количество разрешенных проходов
    //$FFFF – разрешен свободный проход
    function GetEnter(Addr: byte; var Value: word): TModbusError;
    //Состояние выхода
    //0	– проход закрыт;
    //(1-$FFFЕ) – количество разрешенных проходов
    //$FFFF – разрешен свободный проход
    function GetExit(Addr: byte; var Value: word): TModbusError;
    //0 - нормальный режим
    //1 - режим Блокировки
    //2 - режим Антипаники
    function GetPassState(Addr: byte; var Value: word): TModbusError;
    //Ошибки:
    //0 – Ошибок нет
    //не 0 – Наличие ошибок в системе
    function GetErrorState(Addr: byte; var Value: word): TModbusError;
    //Количество проходов в направлении вход
    function GetEnterCount(Addr: byte; var Value: longword): TModbusError;
    //Количество проходов в направлении выход
    function GetExitCount(Addr: byte; var Value: longword): TModbusError;
    //Состояние входа
    //0 - Проход в направлении вход свободен
    //Не 0 – позиция(наличие) объекта с направления вход
    function GetEnterState(Addr: byte; var Value: word): TModbusError;
    //Состояние выхода
    //0 - Проход в направлении выход свободен
    //Не 0 – позиция(наличие) объекта с направления выход
    function GetExitState(Addr: byte; var Value: word): TModbusError;

    //Разрешение накопления проходов
    //0 – Накопление проходов запрещено
    //1 - Накопление проходов разрешено
    function SetAccumulPas(Addr: byte; State: byte): TModbusError;
    //0 - Установить нормальный режим
    //1 - Установить режим Блокировки
    //2 - Установить режим Антипаники
    function SetPassState(Addr: byte; State: byte): TModbusError;
    //скорость обмена между контроллером турникета и контроллером верхнего уровня  деленная на 100
    function SetControllerSpeed(Addr: byte; Value: longword): TModbusError;
    //скорость обмена между контроллером турникета и сканером деленная на 100
    function SetScannerSpeed(Addr: byte; Value: word): TModbusError;
    //0	– проход закрыт;
    //(1-$FFFЕ) – количество разрешенных проходов
    //$FFFF – разрешен свободный проход
    function SetEnter(Addr: byte; State: word): TModbusError;
    //0	– проход закрыт;
    //(1-$FFFЕ) – количество разрешенных проходов
    //$FFFF – разрешен свободный проход
    function SetExit(Addr: byte; State: word): TModbusError;
    //Время необходимое для прохода после разрешения, секунд
    function SetAfterPermissionPassTime(Addr: byte; State: byte): TModbusError;
    //Время необходимое после начала прохода, секунд
    function SetAfterStartPassTime(Addr: byte; State: byte): TModbusError;
    //Таймаут работы двигателя, секунд
    function SetEngineTime(Addr: byte; State: byte): TModbusError;

    //Зеленый Вход: длительность включения сигнала, в 100 мс интервалах (t)
    //длительность выключения сигнала, в 100 мс интервалах (T)
    //Общая длительность цикла индикации, в 100 мс интервалах (time)
    //Значение 0xFFFF – Соответствуют бесконечному времени.
    function SetEnterGreen(Addr: byte; OnTime, OffTime, LengthTime: word): TModbusError;
    //Красный Вход: длительность включения сигнала, в 100 мс интервалах (t)
    //длительность выключения сигнала, в 100 мс интервалах (T)
    //Общая длительность цикла индикации, в 100 мс интервалах (time)
    //Значение 0xFFFF – Соответствуют бесконечному времени.
    function SetEnterRed(Addr: byte; OnTime, OffTime, LengthTime: word): TModbusError;
    //Зеленый Выход: длительность включения сигнала, в 100 мс интервалах (t)
    //длительность выключения сигнала, в 100 мс интервалах (T)
    //Общая длительность цикла индикации, в 100 мс интервалах (time)
    //Значение 0xFFFF – Соответствуют бесконечному времени.
    function SetExitGreen(Addr: byte; OnTime, OffTime, LengthTime: word): TModbusError;
    //Красный Выход: длительность включения сигнала, в 100 мс интервалах (t)
    //длительность выключения сигнала, в 100 мс интервалах (T)
    //Общая длительность цикла индикации, в 100 мс интервалах (time)
    //Значение 0xFFFF – Соответствуют бесконечному времени.
    function SetExitRed(Addr: byte; OnTime, OffTime, LengthTime: word): TModbusError;

    function GetDemo(Addr: byte; var State: word): TModbusError;
    function SetDemo(Addr: byte; State: byte): TModbusError;
  end;

implementation

function TController.GetErrorState(Addr: byte; var Value: word): TModbusError;
begin
  result := ReadData(addr, $48, value);
end;

function TController.GetDemo(Addr:byte; var State: word): TModbusError;
begin
  result := ReadData(addr, $666, state);
end;

function TController.GetEnter(Addr: byte; var Value: word): TModbusError;
begin
  result := ReadData(addr, $40, value);
end;

function TController.GetEnterCard(Addr: byte; var IsNew: boolean;
  var CardNum: TAr; var Count: word): TModbusError;
var
  s: ansistring;
begin
  FillChar(CardNum, SizeOf(CardNum), #0);
  result := ReadData(addr, $100, count);
  IsNew := Boolean(hi(count));
  count := lo(count);

  if (count <> 0) and IsNew then
  begin
    result := ReadData(addr, $101, s, round(count / 2));
    if (count = 15)  and (Byte(s[14]) = $0D) and (Byte(s[15]) = $0A) then
      count := count - 2; // -$0D$0A
    move(s[1], CardNum[0], count);
  end
end;

function TController.GetEnterCount(Addr: byte; var Value: longword): TModbusError;
var
  n: word;
begin
  ReadData(addr, $45, n);
  value := n;
  result := ReadData(1, $44, n);
  value := value shl 16 + n;
end;

function TController.GetEnterState(Addr: byte; var Value: word): TModbusError;
begin
  result := ReadData(addr, $42, value);
end;

function TController.GetExit(Addr: byte; var Value: word): TModbusError;
begin
  result := ReadData(addr, $41, value);
end;

function TController.GetExitCard(Addr: byte; var IsNew: boolean;
  var CardNum: TAr; var Count: word): TModbusError;
var
  s: ansistring;
begin
  FillChar(CardNum, SizeOf(CardNum), #0);
  result := ReadData(addr, $200, count);
  IsNew := Boolean(hi(count));
  count := lo(count);

  if (count <> 0) and IsNew then
  begin
    result := ReadData(addr, $201, s, round(count / 2));
    if (count = 15)  and (Byte(s[14]) = $0D) and (Byte(s[15]) = $0A) then
      count := count - 2; // -$0D$0A
    move(s[1], CardNum[0], count);
  end
end;

function TController.GetExitCount(Addr: byte; var Value: longword): TModbusError;
var
  n: word;
begin
  ReadData(addr, $47, n);
  value := n;
  result := ReadData(addr, $46, n);
  value := value shl 16 + n;
end;

function TController.GetExitState(Addr: byte; var Value: word): TModbusError;
begin
  result := ReadData(addr, $43, value);
end;

function TController.GetPassState(Addr: byte; var Value: word): TModbusError;
begin
  result := ReadData(addr, $3000, value);
end;

function TController.ReadData(Addr: byte; RegAddr: word; var Ans: ansistring; Count: byte = 1): TModbusError;
var
  param: TAddrQtyRec;
  output: TResultRec;
begin
  ans := '';
  param.Addr := RegAddr;
  param.Qty := Count;

  Result := FModbus.ReadHoldingRegisters(Addr, param, output);

  if Result = merNone then
  begin
    if output.Qty <> 0 then
    begin
      SetLength(ans, output.Qty);
      move(output.data[1], ans[1], output.Qty);
    end;
  end
end;

function TController.ReadData(Addr: byte; RegAddr: word; var Ans: word): TModbusError;
var
  param: TAddrQtyRec;
  output: TResultRec;
begin
  ans := 0;
  param.Addr := RegAddr;
  param.Qty := 1;
  FillChar(output.data, SizeOf(output.data), #0);

  Result := FModbus.ReadHoldingRegisters(Addr, param, output);

  if Result = merNone then
  begin
    if output.Qty <> 0 then
      ans := output.data[1] shl 8 + output.data[2];
  end
end;

function TController.SetControllerSpeed(Addr: byte; Value: longword): TModbusError;
var
  speed, speed2: longint;
begin
  speed := trunc(value / 100);
  speed2 := not speed;
  result := WriteState(addr, $4310 , [speed, speed2]);
end;

function TController.SetDemo(Addr, State: byte): TModbusError;
begin
  result := WriteState(addr, $666, [state]);
end;

function TController.SetEngineTime(Addr, State: byte): TModbusError;
begin
  result := WriteState(addr, $48, [state]);
end;

function TController.SetEnter(Addr: byte; State: word): TModbusError;
begin
  result := WriteState(addr, $40, [state]);
end;

function TController.SetEnterGreen(Addr: byte; OnTime, OffTime,
  LengthTime: word): TModbusError;
begin
  result := WriteState(addr, $1000, [OnTime, OffTime, LengthTime]);
end;

function TController.SetEnterRed(Addr: byte; OnTime, OffTime,
  LengthTime: word): TModbusError;
begin
  result := WriteState(addr, $1003, [OnTime, OffTime, LengthTime]);
end;

function TController.SetExit(Addr:byte; State: word): TModbusError;
begin
  result := WriteState(addr, $41, [state]);
end;

function TController.SetExitGreen(Addr: byte; OnTime, OffTime,
  LengthTime: word): TModbusError;
begin
  result := WriteState(addr, $1008, [OnTime, OffTime, LengthTime]);
end;

function TController.SetExitRed(Addr: byte; OnTime, OffTime,
  LengthTime: word): TModbusError;
begin
  result := WriteState(addr, $1009, [OnTime, OffTime, LengthTime]);
end;

function TController.SetAccumulPas(Addr, State: byte): TModbusError;
begin
  result := WriteState(addr, $3010, [state]);
end;

function TController.SetAfterPermissionPassTime(Addr,
  State: byte): TModbusError;
begin
  result := WriteState(addr, $3011, [state]);
end;

function TController.SetAfterStartPassTime(Addr, State: byte): TModbusError;
begin
  result := WriteState(addr, $3012, [state]);
end;

function TController.SetPassState(Addr, State: byte): TModbusError;
begin
  result := WriteState(addr, $3000, [state]);
end;

function TController.SetScannerSpeed(Addr: byte; Value: word): TModbusError;
var
  speed: word;
begin
  speed := trunc(value / 100);
  result := WriteState(addr, $4350, [speed]);
end;

function TController.WriteState(Addr: byte; RegAddr: word;
  Value: array of word): TModbusError;
var
  arec: TRegDataRec;
  i, j: byte;
begin
  arec.Addr := RegAddr;

  arec.Qty := length(value);
  j := 1;
  for i := 0 to length(value) - 1 do
  begin
    arec.data[j] := hi(value[i]);
    arec.data[j + 1] := lo(value[i]);
    j := j + 2;
  end;

  Result := FModbus.WriteMultipleRegisters(Addr, arec);
end;

end.

File: TurnstileCPP/uController.cpp
#include "uController.h"


TRoundingMode GRoundingMode = rmUp;

TRoundingMode GetRoundMode()
{
    return GRoundingMode;
}

void SetRoundMode(TRoundingMode roundMode)
{
    GRoundingMode = roundMode;
}

double Dround(double number)
{	
    int32_t Lround;
    
    TRoundingMode rmd;
    double mnumber = number*(-1);

    rmd = GetRoundMode();
    if(rmd==rmNearest)
    {
        Lround = (number >= 0) ? (int32_t)(number + 0.5) : (int32_t)(number - 0.5);
        return (double)Lround;
    }
    else if(rmd==rmDown)
    {
        Lround = (number >= 0) ? (int32_t)(number) : (int32_t)(number - 1);
        return (double)Lround;
    }
    else if(rmd==rmTruncate)
    {
        return (double)((int32_t)(number));
    }
    else // rmd==rmUp or default (undefined)
    {
        if(number > 0)
        {
        Lround = (number - (double)((int32_t)(number)) > 0) ? (int32_t)(number + 1) : (int32_t)(number);
        return (double)Lround;
        }
        else if(number==0)
        {
            return 0;
        }
        else // number < 0
        {
            Lround = (mnumber - (double)((int32_t)(mnumber)) > 0) ? (int32_t)(mnumber + 1) : (int32_t)(mnumber);
            return (double)(Lround*(-1));
        }
    }
}


TModbusError TController :: GetErrorState(byte Addr, Word & Value)
{
    return ReadData(Addr,0x48,Value);
}


TModbusError TController :: GetDemo(byte Addr, Word & State)
{
    return ReadData(Addr,0x666,State);
}


TModbusError TController :: GetEnter(byte Addr, Word & Value)
{
    return ReadData(Addr,0x40,Value);
}


TModbusError TController :: GetEnterCard(byte Addr, bool & IsNew, TAr * CardNum, Word & Count)
{
    bytesArray s;
    TRoundingMode OldRM;
    TModbusError err;

    OldRM = GetRoundMode();
    memset(CardNum->data,0,TARSIZE);
    err = ReadData(Addr, 0x100, Count);
    if(err == merNone)
        {
            IsNew = (bool)((Count & 0xff00)>>8);
            Count = Count & 0x00ff;
            if(Count && IsNew)
            {
                SetRoundMode(rmUp);
                err = ReadData(Addr,0x101,&s,(byte)(Dround(((double)(Count))/2.)));
                if(Count == 15 && s.data[13] == 0x0D && s.data[14] == 0x0A)
                {
                    Count -= 2;  // delete 0x0d 0x0a
                }
                memmove((void *)(CardNum->data),(const void *)(s.data),Count);
            }
        }
    SetRoundMode(OldRM);
    return err;
}


TModbusError TController :: GetEnterCount(byte Addr, LongWord & Value)
{
    Word n;
    TModbusError err;

    ReadData(Addr,0x45,n);
    Value = n;
    err = ReadData(Addr,0x44,n);
    Value = ((Value & 0x0000ffff) << 16) + (LongWord)n;
    return err;
}


TModbusError TController :: GetEnterState(byte Addr, Word & Value)
{
    return ReadData(Addr,0x42,Value);
}


TModbusError TController :: GetExit(byte Addr, Word & Value)
{
    return ReadData(Addr,0x41,Value);
}


TModbusError TController :: GetExitCard(byte Addr, bool & IsNew, TAr * CardNum, Word & Count)
{
    TModbusError err;
    bytesArray s;
    TRoundingMode OldRM;

    OldRM = GetRoundMode();
    memset((void *)CardNum->data,0,TARSIZE);
    err = ReadData(Addr,0x200,Count);
    if(err == merNone)
    {
        IsNew = (bool)((Count & 0xff00) >> 8);
        Count = Count & 0x00ff;
        if(Count && IsNew)
        {
            SetRoundMode(rmUp);
            err = ReadData(Addr,0x201,&s,(byte)(Dround(((double)(Count))/2.)));
            if(Count == 15 && s.data[13] == 0x0D && s.data[14] == 0x0A)
            {
                Count -= 2;  // delete 0x0d 0x0a
            }
            memmove((void *)(CardNum->data),(const void *)(s.data),Count);
        }
    }
    SetRoundMode(OldRM);
    return err;
}


TModbusError TController :: GetExitCount(byte Addr, LongWord & Value)
{
    TModbusError err;
    Word n;

    ReadData(Addr,0x47,n);
    Value = (LongWord)n;
    err = ReadData(Addr,0x46,n);
    Value = ((Value &0x0000ffff)<<16) + (LongWord)n;
    return err;
}


TModbusError TController :: GetExitState(byte Addr, Word & Value)
{
    return ReadData(Addr,0x43,Value);
}


TModbusError TController :: GetPassState(byte Addr, Word & Value)
{
    return ReadData(Addr,0x3000,Value);
}


TModbusError TController :: ReadData(byte Addr, Word RegAddr, bytesArray * Ans, byte Count = 1)
{
	TAddrQtyRec param;
	TResultRec output;
	TModbusError err;

	Ans->length = 0;
	param.Addr = RegAddr;
	param.Qty = Count;

	err = FModbus->ReadHoldingRegisters(Addr,&param,&output);
	if(err == merNone)
	{
		if(output.Qty != 0)
		{
			Ans->length = output.Qty;
			memmove((void *)(Ans->data),(const void *)(output.data),output.Qty);
		}
		else
		{
			Ans->length = 0;
			memset((void *)(Ans->data),0,BYTESARRAY_MAXLENGTH+10);
		}
	}
	return err;
}


TModbusError TController :: ReadData(byte Addr, Word RegAddr, Word & Ans)
{
	TAddrQtyRec param;
	TResultRec output;
	TModbusError err;

	Ans = 0;
	param.Addr = RegAddr;
	param.Qty = 1;
	memset((void *)(output.data),0,BYTESARRAY_MAXLENGTH);
	err = FModbus->ReadHoldingRegisters(Addr,&param,&output);
	if(err == merNone)
	{
		if(output.Qty != 0)
		{
			Ans = (((Word)(output.data[0])) << 8) + output.data[1];
		}
	}
	return err;
}

TModbusError TController :: SetControllerSpeed(byte Addr, LongWord Value)
{
    LongWord speed, speed2;
    wordsArray wa;
    speed = Value / 100;
    speed2 = ~speed;
    wa.length = 2;
    wa.data[0] = speed;
    wa.data[1] = speed2;
    return WriteState(Addr,0x4310,&wa);
}


TModbusError TController :: SetDemo(byte Addr, byte State)
{
    wordsArray wa;
    wa.length = 1;
    wa.data[0] = (Word)State;
    return WriteState(Addr,0x666,&wa);
}


TModbusError TController :: SetEngineTime(byte Addr, byte State)
{
    wordsArray wa;
    wa.length = 1;
    wa.data[0] = (Word)State;
    return WriteState(Addr,0x48,&wa);
}


TModbusError TController :: SetEnter(byte Addr, Word State)
{
    wordsArray wa;
    wa.length = 1;
    wa.data[0] = (Word)State;
    return WriteState(Addr,0x40,&wa);
}


TModbusError TController :: SetEnterGreen(byte Addr, Word OnTime, Word OffTime, Word LengthTime)
{
    wordsArray wa;
    wa.length = 3;
    wa.data[0] = OnTime;
    wa.data[1] = OffTime;
    wa.data[2] = LengthTime;
    return WriteState(Addr,0x1000,&wa);
}


TModbusError TController :: SetEnterRed(byte Addr, Word OnTime, Word OffTime, Word LengthTime)
{
    wordsArray wa;
    wa.length = 3;
    wa.data[0] = OnTime;
    wa.data[1] = OffTime;
    wa.data[2] = LengthTime;
    return WriteState(Addr,0x1003,&wa);
}


TModbusError TController :: SetExit(byte Addr, Word State)
{
    wordsArray wa;
    wa.length = 1;
    wa.data[0] = (Word)State;
    return WriteState(Addr,0x41,&wa);
}


TModbusError TController :: SetExitGreen(byte Addr, Word OnTime, Word OffTime, Word LengthTime)
{
    wordsArray wa;
    wa.length = 3;
    wa.data[0] = OnTime;
    wa.data[1] = OffTime;
    wa.data[2] = LengthTime;
    return WriteState(Addr,0x1008,&wa);
}


TModbusError TController :: SetExitRed(byte Addr, Word OnTime, Word OffTime, Word LengthTime)
{
    wordsArray wa;
    wa.length = 3;
    wa.data[0] = OnTime;
    wa.data[1] = OffTime;
    wa.data[2] = LengthTime;
    return WriteState(Addr,0x1009,&wa);
}


TModbusError TController :: SetAccumulPas(byte Addr, byte State)
{
    wordsArray wa;
    wa.length = 1;
    wa.data[0] = (Word)State;
    return WriteState(Addr,0x3010,&wa);
}


TModbusError TController :: SetAfterPermissionPassTime(byte Addr, byte State)
{
    wordsArray wa;
    wa.length = 1;
    wa.data[0] = (Word)State;
    return WriteState(Addr,0x3011,&wa);
}


TModbusError TController :: SetAfterStartPassTime(byte Addr, byte State)
{
    wordsArray wa;
    wa.length = 1;
    wa.data[0] = (Word)State;
    return WriteState(Addr,0x3012,&wa);
}


TModbusError TController :: SetPassState(byte Addr, byte State)
{
    wordsArray wa;
    wa.length = 1;
    wa.data[0] = (Word)State;
    return WriteState(Addr,0x3000,&wa);
}


TModbusError TController :: SetScannerSpeed(byte Addr, Word Value)
{
    Word speed;
    wordsArray wa;

    speed = Value / 100;
    wa.length = 1;
    wa.data[0] = (Word)speed;
    return WriteState(Addr,0x4350,&wa);
}

TModbusError TController :: WriteState(byte Addr, Word RegAddr, wordsArray * Value)
{
 	TRegDataRec arec;
 	byte i, j;
 	arec.Addr = RegAddr;
 	arec.Qty = (Value->length);    //*2;  // !!!!!
 	j = 0;
 	for(i=0; i<(Value->length); i++)
 	{
 		arec.data[j] = (byte)(((Value->data[i]) & 0xff00) >> 8);
 		arec.data[j+1] = (byte)((Value->data[i])&0x00ff);
 		j+=2;
 	}
	return FModbus->WriteMultipleRegisters(Addr, &arec);
}

